PurchasesCoreSwift/Purchasing/PurchasesOrchestrator.swift

```diff
@@ -0,0 +1,506 @@
+//
+//  Copyright RevenueCat Inc. All Rights Reserved.
+//
+//  Licensed under the MIT License (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      https://opensource.org/licenses/MIT
+//
+//  PurchasesOrchestrator.swift
+//
+//  Created by Andrés Boedo on 10/8/21.
+
+import Foundation
+import StoreKit
+
+public typealias PurchaseCompletedBlock = (SKPaymentTransaction?, PurchaserInfo?, Error?, Bool) -> Void
+public typealias RCDeferredPromotionalPurchaseBlock = (@escaping PurchaseCompletedBlock) -> Void
+
+@objc(RCPurchasesOrchestratorDelegate) public protocol PurchasesOrchestratorDelegate {
+    func shouldPurchasePromoProduct(_ product: SKProduct, defermentBlock: @escaping RCDeferredPromotionalPurchaseBlock)
+}
+
+// Todo(post-migration): make internal
+@objc(RCPurchasesOrchestrator) public class PurchasesOrchestrator: NSObject {
+
+    @objc public var finishTransactions: Bool { systemInfo.finishTransactions }
+    @objc public var allowSharingAppStoreAccount: Bool {
+        get {
+            if let allow = _allowSharingAppStoreAccount {
+                return allow
+            } else {
+                return identityManager.currentUserIsAnonymous
+            }
+        }
+        set {
+            _allowSharingAppStoreAccount = newValue
+        }
+    }
+
+    @objc public weak var maybeDelegate: PurchasesOrchestratorDelegate?
+
+    private var _allowSharingAppStoreAccount: Bool?
+    private var presentedOfferingIDsByProductID: [String: String] = [:]
+    private var purchaseCompleteCallbacksByProductID: [String: PurchaseCompletedBlock] = [:]
+
+    // todo: remove explicit unwrap once nullability in identityManager is updated
+    private var appUserID: String { identityManager.maybeCurrentAppUserID! }
+    private var unsyncedAttributes: SubscriberAttributeDict {
+        subscriberAttributesManager.unsyncedAttributesByKey(appUserID: self.appUserID)
```

Comment:
curious about the decision to pull subscriberAttributesManager into this class? it doesn't feel directly purchase-related to me

Reply:
Unfortunately, we need it to get access to `unsyncedAttributes`. That is used in: `syncPurchases(receiptRefreshPolicy` `postReceipt(withTransaction`

Reply:
thoughts on maybe having the subscriberAttributesManager work as part of the `PurchasesOrchestratorDelegate` to avoid the need to pass it and to further separate things?

Reply:
I think that's potentially a clever way to accomplish this. The downside that I can see is that the delegate is optional, _and_ it isn't passed in as part of the initialization of the object. So, if the delegate doesn't get set or goes away, that could present a problem. Currently we treat the delegate as truly optional by not crashing when it's missing:
```swift
public func storeKitWrapper(_ storeKitWrapper: StoreKitWrapper,
                                shouldAddStorePayment payment: SKPayment,
                                for product: SKProduct) -> Bool {
        productsManager.cacheProduct(product)

        guard let delegate = maybeDelegate else { return false }
```

Instead of being ok with no delegate, we could:
- require it as part of the initializer- the property would still be `optional` in order to satisfy the `weak`ness of the reference. Buuutttt then we:
- change the accessor so that if it disappears, we `fatalError`.

There might be other things we could do, but we'll always need to have that delegate be optional, and so we can't really enforce it being non-nil unless we crash moving the error from compile-time to runtime.

This is a longwinded way of saying, I think that solution is clever, but in this instance, it might still be better to pass in the `subscriberAttributesManager`

In the future we should definitely consider what we can do to refactor this and make it a lot nice as injecting `11` dependencies is kind of a lot.

Reply:
Ah, I see how that could become a bigger refactor...perhaps we have a separate delegate for subscriber attribute things. But this kinda discussion might be needed for lots of classes, and definitely not something we need to do during migration.